---
output:
  knitrBootstrap::bootstrap_document:
    theme: readable
    highlight: zenburn
    theme.chooser: TRUE
    highlight.chooser: TRUE
  html_document:
    toc: true
    highlight: zenburn
---

```{r setup, echo=FALSE}
knitr::opts_chunk$set(tidy=TRUE, highlight=TRUE, dev="png",
               cache=FALSE, highlight=TRUE, autodep=TRUE, warning=FALSE, error=FALSE,
               message=FALSE, prompt=TRUE, comment='', fig.cap='', 
               base.dir="~/cache/johnson_rnaseq/results")
```

# Overview

# Load read counts and summary information

```{r library-load}
library(ggplot2)
library(reshape)
library(gplots)
library(edgeR)
library(CHBUtils)
library(vsn)
project_summary = "project-summary.csv"
counts_file = "combined.counts"
cbPalette <- c("#999999", "#E69F00", "#56B4E9", "#009E73", "#F0E442",
"#0072B2", "#D55E00", "#CC79A7")
summarydata = read.table(project_summary, header=TRUE, sep=",")
rownames(summarydata) = summarydata$Name
summarydata = summarydata[order(rownames(summarydata)),]
counts = read.table(counts_file, header=TRUE, row.names="id")
counts = counts[, order(colnames(counts))]
```


# Quality control metrics

## Mapped reads
```{r mapped-plot}
ggplot(summarydata, aes(x=Name, y=Mapped)) +
    geom_bar(stat="identity") +
    theme(axis.text.x = element_text(angle = 90)) +
    ylab("mapped reads") + xlab("")
```

Some of the samples are super deep, 100 million reads is about four times the necessary
depth for doing DGE-- more money could be pumped into doing replicates for future experiments.


## Genomic mapping rate
```{r mapping-rate-plot}
ggplot(summarydata, aes(x=Name, y=Mapping.Rate)) +
    geom_bar(stat="identity") +
    theme(axis.text.x = element_text(angle = 90)) +
    ylab("mapping rate") + xlab("")
```

We can map about 90-95% of the reads, this is a good amount.

## Number of genes detected
```{r genes-detected-plot}
ggplot(summarydata, aes(x=Name, y=Genes.Detected)) +
    geom_bar(stat="identity") +
    theme(axis.text.x = element_text(angle = 90)) +
    ylab("genes detected") + xlab("")
```

Number of genes detected is a rough measure of the complexity of the library. This
correlates with read depth, as you sequence deeper you can pick up rare one-off
copies of genes.

## Exonic mapping rate
```{r exonic-mapping-plot}
ggplot(summarydata, aes(x=Name, y=Exonic.Rate)) +
    geom_bar(stat="identity") +
    theme(axis.text.x = element_text(angle = 90)) +
    ylab("exonic mapping rate") + xlab("")
```

Most of the reads that align map to exons, another indicator of a good quality RNA-seq
experiment.

## rRNA mapping rate
```{r rRNA-rate-plot}
ggplot(summarydata, aes(x=Name, y=rRNA.rate)) +
    geom_bar(stat="identity") +
    theme(axis.text.x = element_text(angle = 90)) +
    ylab("rRNA rate") + xlab("")
```

## Estimated fragment length of paired-end reads
```{r fragment-length-plot}
ggplot(summarydata, aes(x=Name, y=Fragment.Length.Mean)) +
    geom_bar(stat="identity") +
    theme(axis.text.x = element_text(angle = 90)) +
    ylab("fragment length") + xlab("")
```

There is some variation to the average insert size of the genes, usually we would expect
these to pretty close to each other.

## Boxplot of raw counts per gene
```{r boxplot-raw}
melted = melt(counts)
colnames(melted) = c("sample", "count")
melted$sample = factor(melted$sample)
melted$sample = reorder(melted$sample, colnames(counts))
melted$count = log(melted$count)
ggplot(melted, aes(x=sample, y=count)) + geom_boxplot() +
  theme(axis.text.x = element_text(angle = 90, hjust = 1)) + xlab("")
```

## Correlation heatmap of raw counts
```{r heatmap-raw}
p = ggheatmap(cor(counts))
ggheatmap.show(p)
```


## Boxplot of TMM-normalized counts per gene
Trimmed mean of M-values (TMM) normalization is described
[here](http://genomebiology.com/2010/11/3/R25)

Robinson, M. D., & Oshlack, A. (2010). A scaling normalization method for differential expression analysis of RNA-seq data. Genome Biology, 11(3). doi:10.1186/gb-2010-11-3-r25

```{r boxplot-normalized}
y = DGEList(counts=counts)
y = calcNormFactors(y)
normalized_counts = cpm(y, normalized.lib.sizes=TRUE)
melted = melt(normalized_counts)
colnames(melted) = c("gene", "sample", "count")
melted$sample = factor(melted$sample)
melted$sample = reorder(melted$sample, colnames(counts))
melted$count = log(melted$count)
ggplot(melted, aes(x=sample, y=count)) + geom_boxplot() +
  theme(axis.text.x = element_text(angle = 90, hjust = 1)) + xlab("")
```

## Correlation heatmap of TMM-normalized counts
```{r heatmap-normalized}
p = ggheatmap(cor(normalized_counts))
ggheatmap.show(p)
```

Even without normalizing, the distributions of counts look similar. Heatmaps of the
Pearson correlation between samples show some separation based on whether or not the
samples included the third FACS antibody.

## Correlation (Spearman) heatmap of TMM-normalized counts
```{r spearman-heatmap-normalized}
p = ggheatmap(cor(normalized_counts, method="spearman"))
ggheatmap.show(p)
```
The Spearman correlation shows clustering based on the replicate. Depending on how
the samples were prepared, it might be useful  to block on replicate in the model and
do pairwise comparisons across the three conditions within each replicate.

## MDS plot of TMM-normalized counts
```{r mds-normalized}
mds(normalized_counts, k=length(colnames(normalized_counts)) - 1)
```

MDS is more resistant to outliers than the Pearson correlation, here
you can also see a separation based on the positive/negative status of
the samples.

## Heatmap of top 30 most expressed genes
```{r top-count-genes, results='asis'}
select = order(rowMeans(counts),decreasing=TRUE)[1:30]
p = ggheatmap(as.matrix(counts[select,]))
ggheatmap.show(p)
```

# Differential expression
```{r deseq2-expression-analysis, results='asis'}
library(DESeq2)
dds = DESeqDataSetFromMatrix(countData=counts,
    colData=summarydata, design = ~ replicate + condition)
dds = DESeq(dds)
contrasts = combn(levels(summarydata$condition), 2, simplify=FALSE)
```

## Effect of variance stabilization
```{r deseq-diagnostics, results='asis'}
par(mfrow=c(1,3))
notAllZero <- (rowSums(counts(dds))>0)
rld <- rlog(dds)
vsd <- varianceStabilizingTransformation(dds)
rlogMat <- assay(rld)
vstMat <- assay(vsd)

meanSdPlot(log2(counts(dds,normalized=TRUE)[notAllZero,] + 1),
           ylim = c(0,2.5))
meanSdPlot(assay(rld[notAllZero,]), ylim = c(0,2.5))
meanSdPlot(assay(vsd[notAllZero,]), ylim = c(0,2.5))
```

## Dispersion estimates
```{r dispersion-estimate}
plotDispEsts(dds)
```



```{r deseq2-handler}
handle_deseq2 = function(dds, summarydata, column) {
  all_combs = combn(levels(summarydata[,column]), 2, simplify=FALSE)
  all_results = list()
  contrast_strings = list()
  for(comb in all_combs) {
    contrast_string = paste(comb, collapse=" vs ")
    contrast = c(column, comb)
    res = results(dds, contrast=contrast)
    res = res[order(res$padj),]
    all_results = c(all_results, res)
    contrast_strings = c(contrast_strings, contrast_string)
  }
  names(all_results) = contrast_strings
  return(all_results)
}
```

```{r annotate-results}
dataset = "hsapiens_gene_ensembl"
filter = "ensembl_gene_id"
symbol = "hgnc_symbol"

annotate_df = function(df, dataset, filter, symbol) {
  require(biomaRt)
  ensembl = useMart('ensembl', dataset = dataset)
  annot.df = getBM(attributes=c(filter, "gene_biotype", symbol), filters=c(filter),
    values=rownames(df), mart=ensembl)
  m = merge(df, annot.df, by.x="row.names", by.y=filter)
  colnames(m)[1] = "id"
  return(m)
}
```

## MA-plots
  
```{r DESeq-output, results='asis'}
all_results = handle_deseq2(dds, summarydata, "condition")
for(i in seq(length(all_results))) {
  plotMA(all_results[[i]])
  title(paste("MA plot for contrast", names(all_results)[i]))
}
```

## Differentially expressed genes
```{r DESeq-tables, results='asis'}
for(i in seq(length(all_results))) {
  out_df = annotate_df(data.frame(all_results[[i]]), dataset, filter, symbol)
  out_df = out_df[order(out_df$padj),]
  cat(paste("Number of genes with FDR < 0.1:", sum(out_df$padj < 0.1, na.rm=TRUE)))
  cat("\n")
  cat(paste("Lowest adjusted p-value hits for", names(all_results)[i]))
  cat("\n")
  knitr::kable(head(out_df))
  write.table(out_df, file=paste(names(all_results)[i], ".tsv", sep=""),
                         quote=FALSE, sep="\t", row.names=FALSE, col.names=TRUE)
  cat("\n")
}
save.image("de.RData")
```
